### ISSUE TITLE
[Feature Request] Local multiplayer

### ISSUE THREAD

**Issue by yuripourre** [diasurgical/devilutionX#1237]
Would be awesome if we could play devilutionX as local multiplayer.

I have a roadmap here that I will try to implement:

- [ ] Abstract gamepad
- [ ] Handle gamepads using index (player 1 = 0, player 2 = 1, etc.)
- [ ] Allows simple joysticks (instead of gamepads) I have something ready, will share later
- [ ] Modify handler classes to handle multiple players
- [ ] Modify UI to show multiple HUDs
- [ ] Modify multiplayer to accommodate a mix of local and remote players

**Comment by Chance4us** [comment]
Do you mean multiplayer in a split screen?

**Comment by yuripourre** [comment]
@Chance4us I was thinking about a single screen but I can think about it.

**Comment by AJenbo** [comment]
You can actually start multiple windows, don't know if that will be helpful in any way but it would be another way to go about it

**Comment by galaxyhaxz** [comment]
Couch co-op would be sure nice. Of course you'd have to make it a lot more like the PSX so the game would pause when bringing up inventory and stuff, it could be a hassle.

**Comment by NikoVP** [comment]
A separate gui should be created for local coop. With modern resolutions partial covering of the screen shouldn't be much of a problem.

**Comment by AJenbo** [comment]
Just require 1280x480 to run :D

**Comment by julealgon** [comment]
What about the camera? I see no tasks related to it. The standard camera which is centered on one player will not cut it.

**Comment by galaxyhaxz** [comment]
Yes, you will have to do what the PSX did and give the players a "leash" where they can't walk further until all other players are within the range. Additionally if one player teleports it must work for all.

**Comment by yuripourre** [comment]
> What about the camera? I see no tasks related to it. The standard camera which is centered on one player will not cut it.

@julealgon
Good point, added it!

> Yes, you will have to do what the PSX did and give the players a "leash" where they can't walk further until all other players are within the range. Additionally if one player teleports it must work for all.

@galaxyhaxz 
That's what I have I mind. Thanks for mentioning teleports, it will be hard. Now, I am concerned about how annoying would be the camera changing too fast because of teleports/phasing.

**Comment by AJenbo** [comment]
I think the PSX camera was on a spring and not always dead center between the players.

**Comment by julealgon** [comment]
> Yep the camera on PSX jittered and lagged behind

I'm not a fan of the standard camera exactly because it lagged behind... I'd actually prefer something which is basically the opposite: make the camera look ahead in the direction you are facing.

That would work very well on multiplayer though (or would be harder to implement I guess).

**Comment by AJenbo** [comment]
> That would work very well on multiplayer though (or would be harder to implement I guess).

wouldn't?

**Comment by yuripourre** [comment]
Proposed UI using original graphics:
![minimal-ui](https://user-images.githubusercontent.com/1824878/113489497-fa4a9b80-949a-11eb-8da2-1e9bb5c56edd.jpg)

**Comment by julealgon** [comment]
Showing the face is a pretty cool idea, even though maybe it would still be worth it showing the player number somewhere, since multiple players can have the same class.

Maybe a "PX" text at the bottom right corner of the image would make sense?

**Comment by AJenbo** [comment]
Or the hero name?

**Comment by yuripourre** [comment]
@julealgon This idea might work well for the level, some games show that over the portrait. I am afraid that player number is not clear enough to identify the player (maybe if player had a marker close to the character).

@AJenbo yeah, I am trying to figure out where to include the name, having the name might work better to identify the player indeed.

**Comment by qndel** [comment]
> @julealgon This idea might work well for the level, some games show that over the portrait. I am afraid that player number is not clear enough to identify the player (maybe if player had a marker close to the character).
> 
> @AJenbo yeah, I am trying to figure out where to include the name, having the name might work better to identify the player indeed.

Seeing the actual player number somewhere might be helpful because monsters will target the player with lower number

**Comment by yuripourre** [comment]
Another attempt after more tweaks. I tried some combinations of ideas. What do you think?

![coop-minimal-ui-v3(fw)](https://user-images.githubusercontent.com/1824878/113519989-575f5380-9566-11eb-86ed-7a6315793518.jpg)

**Comment by AJenbo** [comment]
Personally, I would go for fewer gold borders, and the PvP mode is global so 4 buttons don't actually make that much sense.

**Comment by julealgon** [comment]
Would like to see a design that:
- Fits in 640x480
- Shows main elements of the panel (xp bar and durability icons are missing here)
- Ideally showed the 4 hotkeys for each player (or showed additional UI so that players could hotkey their spells somehow - not clear how that would be done here)
- Is a bit more similar to other UI elements in the game (see ajenbo's comment)
- Showed the P1/P2/P3/P4 text in the same color used to draw the player arrows in the automap

**Comment by AJenbo** [comment]
> Showed the P1/P2/P3/P4 text in the same color used to draw the player arrows in the automap

Could have a border with the same color as the arrows on the map

**Comment by yuripourre** [comment]
One more attempt (640x480)
![minimalistic_flip](https://user-images.githubusercontent.com/1824878/113523865-17599a00-9581-11eb-8c94-776b8c6aa773.jpg)

**Comment by julealgon** [comment]
What's the worst-case scenario for the UI? Can you post a screenshot with 
- 4 players
- the biggest possible character name allowed in the game currently in terms of length
- with all gear pieces breaking (helm, armor, sword, shield) 
- how it looks like when you have 20+ spells learned and is selecting hotkeys.
- with inventory and character screens open
- with a scroll hotkey, showing how many scrolls are left for the selected spell

**Comment by AJenbo** [comment]
Only allowing 2 players like the original PSX version wouldn't be the end of the world either.

**Comment by julealgon** [comment]
> Only allowing 2 players like the original PSX version wouldn't be the end of the world either.

I don't disagree. I just mentioned it because Yuri himself was targeting 4 players so I wanted to make sure the UI could handle it.

**Comment by AJenbo** [comment]
Wasn't strictly a reply to you either :)

**Comment by jan-kleks** [comment]
It might be worth looking at how other action RPGs implement local multi such as Diablo 4 to search for ideas.

**Comment by yuripourre** [comment]
That's the best I could think so far. I am working on the couch co-op, I had to change some of the joystick buttons, there are still a lot of bugs but it's progressing.

<img width="1920" height="1200" alt="Image" src="https://github.com/user-attachments/assets/20a96e86-cbe3-4ce0-81be-812f6502b06e" />

<img width="1920" height="1200" alt="Image" src="https://github.com/user-attachments/assets/5b199d6a-9b39-4511-8840-7ae51b41bd10" />

**Comment by yuripourre** [comment]
![Image](https://github.com/user-attachments/assets/0eda2b24-8ce5-48b3-bd53-c08e7b2612be)

**Comment by yuripourre** [comment]
That's it, PR updated and rebased, ready for review.

https://github.com/diasurgical/DevilutionX/pull/8354
